clarity/src/vm/types/signatures.rs

```diff
@@ -873,6 +873,48 @@ impl TypeSignature {
             _ => Ok(self.clone()),
         }
     }
+
+    /// Goes recursively through a type to [concretize](TypeSignature::concretize)
+    /// the inner [ListUnionType] and [CallableType] variants.
+    pub fn concretize_deep(self) -> Result<Self> {
+        match self {
+            TypeSignature::NoType
+            | TypeSignature::IntType
+            | TypeSignature::UIntType
+            | TypeSignature::BoolType
+            | TypeSignature::PrincipalType
+            | TypeSignature::SequenceType(SequenceSubtype::BufferType(_))
+            | TypeSignature::SequenceType(SequenceSubtype::StringType(_))
+            | TypeSignature::TraitReferenceType(_) => Ok(self),
+            TypeSignature::OptionalType(opt_ty) => Ok(TypeSignature::OptionalType(Box::new(
+                opt_ty.concretize_deep()?,
+            ))),
+            TypeSignature::ResponseType(resp_ty) => {
+                let (ok_ty, err_ty) = *resp_ty;
+                Ok(TypeSignature::ResponseType(Box::new((
+                    ok_ty.concretize_deep()?,
+                    err_ty.concretize_deep()?,
+                ))))
+            }
+            TypeSignature::SequenceType(SequenceSubtype::ListType(ListTypeData {
+                max_len,
+                entry_type,
+            })) => Ok(TypeSignature::SequenceType(SequenceSubtype::ListType(
+                ListTypeData {
+                    max_len,
+                    entry_type: Box::new(entry_type.concretize_deep()?),
+                },
+            ))),
+            TypeSignature::TupleType(TupleTypeSignature { type_map }) => {
+                let mut type_map_ref = type_map.as_ref().clone();
```

Comment:
The `type_map` type change from `HashMap<>` to `Arc<BtreeMap>` 
I'm not sure about this `as_ref().clone()` that I had to do